lib/Basics/GlobalResourceMonitor.cpp

```diff
@@ -0,0 +1,88 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2021 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Jan Steemann
+////////////////////////////////////////////////////////////////////////////////
+
+#include "Basics/GlobalResourceMonitor.h"
+#include "Basics/debugging.h"
+#include "Basics/system-compiler.h"
+#include "Logger/LogMacros.h"
+
+namespace {
+// a global shared instance for tracking all resources
+arangodb::GlobalResourceMonitor instance;
+}
+
+namespace arangodb {
+
+/// @brief set the global memory limit
+void GlobalResourceMonitor::memoryLimit(std::int64_t value) noexcept {
+  TRI_ASSERT(value >= 0);
+  _limit = value;
+}
+  
+/// @brief return the global memory limit
+std::int64_t GlobalResourceMonitor::memoryLimit() const noexcept {
+  return _limit;
+}
+  
+/// @brief return the current global memory usage
+std::int64_t GlobalResourceMonitor::current() const noexcept {
+  return _current.load(std::memory_order_relaxed);
+}
+
+/// @brief the current memory usage to zero
+void GlobalResourceMonitor::clear() noexcept {
+  _current = 0;
+}
+  
+/// @brief increase global memory usage by <value> bytes. if increasing exceeds the
+/// memory limit, does not perform the increase and returns false. if increasing
+/// succeds, the global value is modified and true is returned
+bool GlobalResourceMonitor::increaseMemoryUsage(std::int64_t value) noexcept {
+  TRI_ASSERT(value >= 0);
+  auto previous = _current.fetch_add(value, std::memory_order_relaxed);
+  if (_limit > 0 && ADB_UNLIKELY(previous + value > _limit)) {
+    // the allocation would exceed the global maximum value, so we need to roll back.
+    // revert the change to the global counter
+    _current.fetch_sub(value, std::memory_order_relaxed);
+    return false;
+  }
```

Comment:
Alternatively we could perform this increment with a CAS-loop. This would have the advantage that we would recognize a threshold exhaustion _before_ we make the change, so there is no need for a rollback. That way we could also avoid the correction logic in ResourceUsage.cpp.

However, under high contention a CAS-loop is less efficient than a fetch-add as threads have to perform retries (at least on x86; on ARM fetch-add is implemented as a CAS-loop anyway).

Reply:
Performance tests showed no relevant performance difference between the CAS loop and fetch-add implementation. But unfortunately we still need the correction logic since this issue is caused by the changes to the _local_ counter. However, we still decided to go with the CAS loop because this way we can at least avoid the temporary limit overflow.